da-contracts/contracts/BlobsL1DAValidatorZKsyncOS.sol

```diff
@@ -0,0 +1,99 @@
+            publishedBlobs[versionedHash] = block.number;
+            ++versionedHashIndex;
+        }
+    }
+
+    function isBlobAvailable(bytes32 _versionedHash) public view returns (bool) {
+        uint256 blockOfPublishing = publishedBlobs[_versionedHash];
+
+        // While `block.number` on all used L1 networks is much higher than `BLOB_EXPIRATION_BLOCKS`,
+        // we still check that `blockOfPublishing > 0` just in case.
+        return blockOfPublishing > 0 && block.number - blockOfPublishing <= BLOB_EXPIRATION_BLOCKS;
+    }
+
+    /// @inheritdoc IL1DAValidator
+    function checkDA(
+        uint256, // _chainId
+        uint256, // _batchNumber
+        bytes32 _l2DAValidatorOutputHash,
+        bytes calldata _operatorDAInput,
+        uint256 _maxBlobsSupported
+    ) external view returns (L1DAValidatorOutput memory output) {
+        // As DA input operator should provide concatenated blob versioned hashes of published blobs.
+        // If provided hash equals to 0, blob versioned hash will be taken from blob published with the current transaction.
+        if (_operatorDAInput.length % BLOB_VERSIONED_HASH_SIZE != 0) {
+            revert InvalidBlobsDAInputLength(_operatorDAInput.length);
+        }
+
+        uint256 blobsProvided = _operatorDAInput.length / BLOB_VERSIONED_HASH_SIZE;
+
+        bytes memory publishedVersionedHashes = new bytes(_operatorDAInput.length);
+        uint256 versionedHashIndex = 0;
+
+        for (uint256 index = 0; index < blobsProvided; ++index) {
+            bytes32 versionedHash = bytes32(_operatorDAInput[BLOB_VERSIONED_HASH_SIZE * index : BLOB_VERSIONED_HASH_SIZE * (index + 1)]);
+            if (versionedHash != bytes32(0)) {
+                if (!isBlobAvailable(versionedHash)) {
+                    revert BlobNotPublished();
+                }
+            } else {
+                versionedHash = _getBlobVersionedHash(versionedHashIndex);
+                ++versionedHashIndex;
+            }
+            // efficiently write versionedHash into the byte array
+            assembly {
+                // offset relative to the `publishedVersionedHashes` start
+                // we add 1 to `index` to skip length slot
+                let relativeOffset := mul(add(index, 1), 32)
+                mstore(add(publishedVersionedHashes, relativeOffset), versionedHash)
+            }
+        }
+
+        // verify that published versionedHashes corresponds to what was generated by ZKsync OS for a batch
+        bytes32 publishedVersionedHashesHash = keccak256(publishedVersionedHashes);
+        if(publishedVersionedHashesHash != _l2DAValidatorOutputHash) {
+            revert InvalidBlobsPublished(publishedVersionedHashesHash, _l2DAValidatorOutputHash);
+        }
+
+        // we use this interface for compatability, but in fact outputs not needed for ZKsync OS
+        output.stateDiffHash = bytes32(0);
+        output.blobsLinearHashes = new bytes32[](_maxBlobsSupported);
```

Comment:
maybe just leave the `_maxBlobsSupported` var as unused and spawn empty arrays? this saves gas and also removes the need to have a contract upgrade if the actual number of supported blobs grows on zksync os side